pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,998 @@
+
+	/// Details of a pool.
+	#[pallet::storage]
+	#[pallet::getter(fn pool_data)]
+	pub type PoolData<T: Config> = StorageMap<_, Blake2_128Concat, PoolId<T>, Pool<T::BlockNumber>, ValueQuery>;
+
+	/// Pool balances.
+	#[pallet::storage]
+	#[pallet::getter(fn pool_balances)]
+	pub type PoolBalances<T: Config> =
+		StorageMap<_, Blake2_128Concat, PoolId<T>, (BalanceOf<T>, BalanceOf<T>), ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(<T as Config>::WeightInfo::create_pool())]
+		#[transactional]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			pool_owner: T::AccountId,
+			asset_a: LBPAssetInfo<BalanceOf<T>>,
+			asset_b: LBPAssetInfo<BalanceOf<T>>,
+			sale_duration: (T::BlockNumber, T::BlockNumber),
+			weight_curve: WeightCurveType,
+			pausable: bool,
+		) -> DispatchResultWithPostInfo {
+			T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				!asset_a.amount.is_zero() && !asset_b.amount.is_zero(),
+				Error::<T>::CannotCreatePoolWithZeroLiquidity
+			);
+
+			ensure!(asset_a.id != asset_b.id, Error::<T>::CannotCreatePoolWithSameAssets);
+
+			let asset_pair = AssetPair {
+				asset_in: asset_a.id,
+				asset_out: asset_b.id,
+			};
+
+			ensure!(!Self::exists(asset_pair), Error::<T>::PoolAlreadyExists);
+
+			ensure!(
+				T::MultiCurrency::free_balance(asset_a.id, &pool_owner) >= asset_a.amount,
+				Error::<T>::InsufficientAssetBalance
+			);
+
+			ensure!(
+				T::MultiCurrency::free_balance(asset_b.id, &pool_owner) >= asset_b.amount,
+				Error::<T>::InsufficientAssetBalance
+			);
+
+			let pool_data = Pool::new(asset_a, asset_b, sale_duration, weight_curve, pausable);
+			Self::validate_pool_data(&pool_data)?;
+
+			let deposit = T::PoolDeposit::get();
+
+			T::MultiCurrency::reserve(T::NativeAssetId::get(), &pool_owner, deposit)?;
+			let pool_id = Self::get_pair_id(asset_pair);
+			<PoolDeposit<T>>::insert(&pool_id, &deposit);
+
+			<PoolOwner<T>>::insert(&pool_id, &pool_owner);
+			<PoolAssets<T>>::insert(&pool_id, &(asset_a.id, asset_b.id));
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+
+			T::MultiCurrency::transfer(asset_a.id, &pool_owner, &pool_id, asset_a.amount)?;
+			T::MultiCurrency::transfer(asset_b.id, &pool_owner, &pool_id, asset_b.amount)?;
+
+			<PoolBalances<T>>::insert(&pool_id, &(asset_a.amount, asset_b.amount));
+
+			Self::deposit_event(Event::PoolCreated(
+				pool_owner,
+				asset_a.id,
+				asset_b.id,
+				asset_a.amount,
+				asset_b.amount,
+			));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::update_pool_data())]
+		#[transactional]
+		pub fn update_pool_data(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T>,
+			start: Option<T::BlockNumber>,
+			end: Option<T::BlockNumber>,
+			initial_weights: Option<((AssetId, LBPWeight), (AssetId, LBPWeight))>,
+			final_weights: Option<((AssetId, LBPWeight), (AssetId, LBPWeight))>,
```

Comment:
Perhaps add a check to verify if at least one thing is not None, so there is actually something to update.

Reply:
What to do when all fields are set to None? Throw the PoolUpdated event and skip the rest or throw an error?

Reply:
> 
> 
> What to do when all fields are set to None? Throw the PoolUpdated event and skip the rest or throw an error?

I would just return an error - something saying that there is nothing to update.